rfcs/0005-multi-blockchain-support.md

```diff
@@ -0,0 +1,570 @@
+
+A number of features in the current store interface is based on so-called
+Ethereum block pointers, a pair of block hash and block number. This is changed
+to a generic block pointer type, where the hash is a byte array rather than an
+Ethereum 256-bit hash.
+
+Another part of the current store that is specific to Ethereum is the chain
+store interface. Assuming the ongoing work on chain indexing is developed far
+enough to replace the current block ingestion and block streaming, this
+interface can be removed. The only part that will have to be kept is the
+so-called chain head listener. This is moved into either the chain or the chain
+indexer component.
+
+### Adding Support for a New Blockchain
+
+This section covers what it takes to add support for a new blockchain. Please
+note that this RFC specifies the details only to an extent that allow us to
+assess whether the direction is feasible. Names and data structure details may
+vary in the implementation.
+
+#### Graph Node
+
+Each blockchain integration in Graph Node lives in the `graph-node` repository
+under `network/<name>/`, like the already existing `network/ethereum/` support.
+
+Graph Node includes a number of new component traits and data types to enable
+multi-blockchain support:
+
+```rust
+/// A compact representation of a block on a chain.
+struct BlockPointer {
+    pub number: BigInt,
+    pub hash: Bytes,
+}
+
+/// Represents a block in a particular chain. Each chain has its own
+/// implementation of the `Block` trait.
+trait Block: ToAsc {
+    fn number(&self) -> BigInt;
+    fn hash(&self) -> Bytes;
+    fn pointer(&self) -> BlockPointer;
+    fn parent_pointer(&self) -> Option<BlockPointer>;
+}
+
+/// Represents a decentralized network like Ethereum, Polkadot or similar.
+enum Network {
+    // Variant for blockchain networks.
+    Blockchain(Box<dyn Blockchain>),
+}
+
+/// Represents a blockchain supported by the node.
+trait Blockchain {
+    type Chain: Chain;
+    type Options;
+
+    fn new(options: Self::Options) -> Self;
+    
+    // Looks up a chain by name. Fails if the node doesn't support the chain.
+    fn chain(&self, name: String) -> Result<Self::Chain, Error>;
+}
+
+// Common options for interacting with any chain.
+
+struct LatestBlockOptions {
+    logger: Logger,
+}
+
+struct BlockByNumberOptions {
+    logger: Logger,
+    number: BigInt,
+}
+
+struct BlockByHashOptions {
+    logger: Logger,
+    hash: Bytes
+}
+
+struct BlockStreamOptions<'a> {
+    logger: Logger,
+    data_sources: Vec<&'a DataSource>,
+}
+
+/// Represents a network (essentially an instance of a `Chain`).
+trait Chain {
+    type Block: Block;
+    type ChainIndexer: ChainIndexer<Self::Block>;
+    type BlockStream: BlockStream<Self::Block>;
+
+    // Methods common to all chains. These are needed by the chain indexer
+    // and block stream, potentially also by time-travel queries.
+    async fn latest_block(&self, options: LatestBlockOptions) -> Result<Self::Block, Error>;
```

Comment:
Why not just have the fields of the `Options` struct directly as arguments? If this grows to a large number of arguments, some with default values, then it might make sense to have a separate struct, maybe even as an associated type if it needs to be customized per chain. But that doesn't seem to be the case yet.